pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -225,10 +231,25 @@ func newQueryTripperware(
 			log,
 			engine,
 			limits,
+			cfg.MaxSeriesPerShard,
 			registerer,
 		)
-		queryRangeMiddleware = append(
-			queryRangeMiddleware,
+
+		if cfg.MaxSeriesPerShard > 0 {
```

Comment:
[nit] Since I got tricked while reviewing the PR and other people may get tricked too, I would add a comment to clarify why adding the middleware here is the right place.

```suggestion
                // Inject the cardinality estimation middleware after the time splitting and before query sharding
                // so that we estimate each partial query individually before it gets sharded.
		if cfg.MaxSeriesPerShard > 0 {
```